### DISCUSSION TITLE
When I run the shape_optimiaztion.py with dis_adj transonic_stator_2D it said unknown function name, FLOW_ANGLE_OUT.

### DISCUSSION THREAD

**Discussion by BlackHawk992** [su2code/SU2#2280] [Q&A]
Hello！

I'm a totally beginer currently working on a turbo stator optimization using SU2, I tried to run the testcase to learn how to use, but when I run this [disc_adj_turbomachinery](https://github.com/su2code/SU2/tree/master/TestCases/disc_adj_turbomachinery)
/transonic_stator_2D case, it returned Exception: unknown function name, FLOW_ANGLE_OUT. Please check config_template.cfg for updated list of function names. Have I miss any step or setting before starting? 

I didn't change the cfg file and run with python shape_optimization.py -g DISCRETE_ADJOINT -f .\transonic_stator.cfg

It would be very greatful if you konw where the problem is!
Sorry for can’t solve such a problem myself!

**Comment by joshkellyjak** [comment]
The turbomachinery objective functions and constraints don't currently exist in the latest branch, this is an area of development but has stalled quite considerably recently.

**Reply by xiapxiaoyu1981** [reply]
Dear joshkellyjak,


Similar to the original poster, I am also relatively new to SU2. Through my initial exploration of SU2, I have discovered that the shape_optimization.py script is primarily tailored for optimizing workflows related to wing design. The objective functions and constraints within the script are almost exclusively focused on evaluating wing performance, which is not suitable for turbomachinery.


To incorporate objective functions and constraints specific to turbomachinery, I believe there are two potential approaches:


Without Modifying the Source Code: I have learned that SU2 has a powerful feature called CUSTOM_OBJUNC, which seems to seamlessly integrate into the optimization workflow. This feature allows for flexible construction of the necessary objective functions and constraints using basic variables. However, it is unfortunate that shape_optimization.py does not support the CUSTOM_OBJUNC feature. From various sources, I have learned that FADO might be a good alternative. However, my understanding of FADO is quite limited. Therefore, I would like to ask if FADO can address the lack of turbomachinery-specific objective functions and constraints, as well as the absence of CUSTOM_OBJUNC, in the shape_optimization.py workflow?


Modifying the Source Code: As a beginner in C++ (currently learning), I wonder if it would be possible to add turbomachinery-specific objective functions, such as efficiency, pressure ratio, and entropy loss, to the source code by drawing inspiration from the wing performance metrics (like Drag, etc.).


Sincerely,


Tongtong

**Reply by joshkellyjak** [reply]
CUSTOM_OBJFUNC may be able to use turbomachinery specific objective functions. Implementing the turbo objective functions in C++ should also be fairly simple but there is still ongoing development needed to validate the gradients for turbomachinery simulations (see #2317 ).

**Reply by xiapxiaoyu1981** [reply]
Oh, Josh Kelly Jack, thank you for your reply. I have another question I want to confirm with you. If we use discrete adjoint automatic differentiation to solve the sensitivity of the objective function, do we still need to validate the sensitivity? (At this stage, my understanding of automatic differentiation is that the gradient information is obtained through a mathematical method based on the direct computation of the flow solution)

**Reply by joshkellyjak** [reply]
You should always validate your gradients, both to ensure that what your optimiser _sees is correct_ and as good practice for publication of any results you achieve.

**Reply by xiapxiaoyu1981** [reply]
Hi JoshKellyJak, Thank you very much for your response. I will review some more materials to gain a more comprehensive understanding of this topic.
